### PR TITLE
add GroupSig precompiled contract

### PR DESCRIPTION
**Pull request by Shareong**

### CODE REVIEW

File: libprecompiled/extension/GroupSigPrecompiled.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+class GroupSigPrecompiled : public dev::blockverifier::Precompiled
+{
+public:
+    typedef std::shared_ptr<GroupSigPrecompiled> Ptr;
+    GroupSigPrecompiled();
+    virtual ~GroupSigPrecompiled(){};
+
+    virtual bytes call(dev::blockverifier::ExecutiveContext::Ptr context, bytesConstRef param,
+        Address const& origin = Address());
```
Comment (bxq2011hust) [inline libprecompiled/extension/GroupSigPrecompiled.h]:
```suggestion
        Address const& origin = Address()) override;
```

File: libprecompiled/extension/GroupSigPrecompiled.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+#endif
+
+class GroupSigPrecompiled : public dev::blockverifier::Precompiled
+{
+public:
+    typedef std::shared_ptr<GroupSigPrecompiled> Ptr;
+    GroupSigPrecompiled();
+    virtual ~GroupSigPrecompiled(){};
+
+    virtual bytes call(dev::blockverifier::ExecutiveContext::Ptr context, bytesConstRef param,
```
Comment (bxq2011hust) [inline libprecompiled/extension/GroupSigPrecompiled.h]:
```suggestion
    bytes call(dev::blockverifier::ExecutiveContext::Ptr context, bytesConstRef param,
```

File: libprecompiled/extension/GroupSigPrecompiled.cpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        abi.abiOut(data, signature, message, gpkInfo, paramInfo);
+
+        int valid = 0;
+        int ret = GroupSigApi::group_verify(
+            valid, signature, message, ALGORITHM_METHOD, gpkInfo, paramInfo);
+        if (ret == 0)
+        {
+            if (valid)
+            {
+                result = "true";
```
Comment (bxq2011hust) [inline libprecompiled/extension/GroupSigPrecompiled.cpp]:
```suggestion
                result = true;
```

File: cmake/ProjectGroupSig.cmake [hunk 4]
```diff
@@ -0,0 +1,51 @@
+include(ExternalProject)
+include(GNUInstallDirs)
+
+if (CRYPTO_EXTENSION)
+    ExternalProject_Add(GroupSigLib
+        PREFIX ${CMAKE_SOURCE_DIR}/deps
+        DOWNLOAD_NAME group_sig_lib.tgz
+        DOWNLOAD_NO_PROGRESS 1
+        GIT_REPOSITORY https://github.com/Shareong/GroupSig.git
```
Comment (bxq2011hust) [inline cmake/ProjectGroupSig.cmake]: please use official repo

File: cmake/ProjectGroupSig.cmake [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    set(DEP_INCLUDE_DIR ${SOURCE_DIR}/deps/include)
+    file(MAKE_DIRECTORY ${DEP_INCLUDE_DIR}) 
+
+    find_library(GMP_LIBRARIES NAMES "libgmp.a")
+    find_path(GMP_INCLUDE_DIR "gmp.h")
+    add_library(Gmp UNKNOWN IMPORTED)
+    set_property(TARGET Gmp PROPERTY IMPORTED_LOCATION ${GMP_LIBRARIES})
+    set_property(TARGET Gmp PROPERTY INTERFACE_INCLUDE_DIRECTORIES ${GMP_INCLUDE_DIR})
+
+    add_library(Pbc STATIC IMPORTED)
```
Comment (bxq2011hust) [inline cmake/ProjectGroupSig.cmake]: I don't think we need these.

File: libprecompiled/extension/GroupSigPrecompiled.cpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    std::string result = "false";
+
+    if (func == name2Selector[GroupSig_METHOD_SET_STR])
+    {
+        // groupSigVerify(string)
+        std::string signature, message, gpkInfo, paramInfo;
+        abi.abiOut(data, signature, message, gpkInfo, paramInfo);
+
+        int valid = 0;
+        int ret = GroupSigApi::group_verify(
```
Comment (JimmyShi22) [inline libprecompiled/extension/GroupSigPrecompiled.cpp]: try catch

File: libprecompiled/extension/GroupSigPrecompiled.cpp [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+    dev::eth::ContractABI abi;
+    bytes out;
+    std::string result = "false";
+
+    if (func == name2Selector[GroupSig_METHOD_SET_STR])
+    {
+        // groupSigVerify(string)
+        std::string signature, message, gpkInfo, paramInfo;
+        abi.abiOut(data, signature, message, gpkInfo, paramInfo);
```
Comment (JimmyShi22) [inline libprecompiled/extension/GroupSigPrecompiled.cpp]: try catch

File: libprecompiled/extension/GroupSigPrecompiled.cpp [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+        int valid = 0;
+        int ret = GroupSigApi::group_verify(
+            valid, signature, message, ALGORITHM_METHOD, gpkInfo, paramInfo);
```
Comment (JimmyShi22) [inline libprecompiled/extension/GroupSigPrecompiled.cpp]: return valid and throw if ret != 0

**Pull request merged**